---
title: "cTP-net Vignette"
author: "Zilu Zhou"
date: "06/05/2019"#"`r Sys.Date()`"
abstract: >
 While single cell RNA sequencing (scRNA-seq) is invaluable for studying cell populations, cell-surface proteins are often integral markers of cellular function and serve as primary targets for therapeutic intervention.  Here we propose a transfer learning framework, single *c*ell *T*ranscriptome to *P*rotein prediction with deep neural *net*work (cTP-net), to impute surface protein abundances from scRNA-seq data by learning from existing single-cell multi-omic resources. For more detail, please check our [biorixv preprint, no link yet](www.rstudio.com)
output: 
  rmarkdown::html_document:
    theme: united
    highlight: tango
    toc: true
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
bibliography: DENDRO.bibtex
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# 1. Installation
Install to R/RStudio
Install all packages in the latest version of [R](https://www.r-project.org/).
First, install the supporting Python package ctpnetpy. See the source code of the package [here](http://github.com/zhouzilu/ctpnetpy)

```python
pip install ctpnetpy
```

Next, open R and install the R package cTPnet
```{r, eval=FALSE}
devtools::install_github("zhouzilu/cTPnet")
```


# 2. Questions & issues
If you have any questions or problems when using cTPnet or ctpnetpy, please feel free to open a new issue [here](https://github.com/zhouzilu/cTPnet/issues). You can also email the maintainers of the corresponding packages -- the contact information is shown under Developers & Maintainers.

* [Zilu Zhou](https://statistics.wharton.upenn.edu/profile/zhouzilu/) (zhouzilu at pennmedicine dot upenn dot edu)
  <br>
  Genomics and Computational Biology Graduate Group, UPenn

* [Nancy R. Zhang](https://statistics.wharton.upenn.edu/profile/nzh/) (nzh at wharton dot upenn dot edu)
  <br>
  Department of Statistics, UPenn

# 3. cTP-net analysis pipeline

To accurately impute surface protein abundance from scRNA-seq data, cTP-net employs two steps: (1) denoising of the scRNA-seq count matrix and (2) imputation based on the denoised data through a transcriptome-protein mapping (Figure 1).  The initial denoising, by SAVERX, produces more accurate estimates of the RNA transcript relative abundances for each cell.  Compared to the raw counts, the denoised relative expression values have significantly improved correlation with their cognate protein measurement.  

```{r, out.width = "1000px", fig.align = "center", echo=FALSE}
knitr::include_graphics("https://raw.githubusercontent.com/zhouzilu/cTPnet/master/figure/Pkg_FIG-01.jpg")
```
  **Figure 1**. Overview of cTP-net analysis pipeline, which learns a mapping from the denoised scRNA-seq data to the relative abundance of surface proteins, capturing multi-gene features that reflect the cellular environment and related processes. 

Here we load our demo dataset (11000 genes $\times$ 1000 cells) from Human Cell Atalas.
```{r, message=FALSE, warning=FALSE}
library(cTPnet)
data("cTPnet_demo")
str(demo)
```

where `Info` indicates mutation information such as chromosome, allele nucleitide and position, and `label` indicates the true label, which we don't have in a real data analysis.

### 3.1  

Given $X, N$ and $Z$, DENDRO first apply quality control (QC). In default, DENDRO filter out (1) mutations with less than 5% cells detected (too rare) or more than 95% of cells detected (too common), and (2) cells with total read counts deviated than 5 standard deviation. Check `??FilterCellMutation` for more details. 

```{r, message=FALSE, warning=FALSE}
demo_qc = FilterCellMutation(demo$X,demo$N,demo$Z,demo$Info,demo$label,cut.off.VAF = 0.05, cut.off.sd = 5)
```

The above two plots illustrate two distributions: (left) total number of cells for each mutations and (right) total number of mutations for each cell, with the red line indicating filter thresholds.

```{r, message=FALSE, warning=FALSE}
str(demo_qc)
```


## 3.3 Stage II
### 3.3.1 Genetic divergence matrix calculation

Now DENDRO can calculate the genetic divergence matrix. Check `??DENDRO.dist` for more details.

```{r, message=FALSE, warning=FALSE, fig.width=14}
demo_qc$dist = DENDRO.dist(demo_qc$X,demo_qc$N,demo_qc$Z,show.progress=FALSE)
```

### 3.3.2 Clustering with the genetic divergence matrix

Let's apply hierachical clustering and plot out the clustering result colored by known true label: `demo_qc$clade`. Check `DENDRO.cluster` for more details.

```{r, message=FALSE, warning=FALSE, fig.width=14}
demo_qc$cluster = DENDRO.cluster(demo_qc$dist,label=demo_qc$label)
```

Let's decided the optimal number of clusters using an intra-cluster divergence (icd) measurements.

```{r, message=FALSE, warning=FALSE}
demo_qc$icd = DENDRO.icd(demo_qc$dist,demo_qc$cluster)
demo_qc$optK = 3
demo_qc$DENDRO_label = cutree(demo_qc$cluster,demo_qc$optK)
```

We decide the optimal number of cluster by identifying kink or "elbow point" in the icd plot. In this example, `optK = 3`. It is crucial that if there are multiple "elbow point", the *smallest* one is the most robust.

Let's re-plot our data with DENDRO label

```{r, message=FALSE, warning=FALSE, fig.width=14}
demo_qc$cluster = DENDRO.cluster(demo_qc$dist,label=demo_qc$DENDRO_label)
```

### 3.3.3 Re-estimate mutation profile within each cluster and QC

DENDRO further re-esimate the subclone-level mutation profiles by pooling all reads within each reads together with a maximum liklihood approach [@li2012likelihood]. Check `??DENDRO.recalculate` for more details.

```{r, message=FALSE, warning=FALSE}
demo_cluster = DENDRO.recalculate(demo_qc$X,demo_qc$N, demo_qc$Info, demo_qc$DENDRO_label, cluster.name=c('Cluster3','Cluster2','Cluster1'))
```
`cluster.name` specifies the cluster name given the clustering order (1, 2, 3, ...).

## 3.4 Stage III
### 3.4.1 Evolutionary tree construction

Given the filtered cluster-level mutation profiles, we now can construct an neighbor-joining tree using algorithm implemented in package `phangorn`. See `??DENDRO.tree` for more details.

```{r, message=FALSE, warning=FALSE}
DENDRO.tree(demo_cluster$Z)
```

In this phylogenetic tree, Cluster1 has greater genetic divergence compared with Cluster2 and Cluster3, which is consistent with our data generating process. 

### 3.4.2 Other analysis

User could further perform joint differential expression analysis and differential mutation analysis between different subclone groups. Mutation profile across clones is sored at `demo_cluster$Z`.

Differential expression analysis packages are wide-spread. Two methods that I personally preferred are [Seurat MAST implementation](https://satijalab.org/seurat/get_started.html) [@seurat2018] and [scDD](https://bioconductor.org/packages/release/bioc/html/scDD.html) [@scDD2016].

Gene set enrichment analysis is available at [MSigDB, Broad Institute](http://software.broadinstitute.org/gsea/msigdb/) [@gsea2005].


# 4. Session info

```{r sessionInfo}
sessionInfo()
```

# 5. References

